

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>supremo_lite.chromosome_utils &mdash; supremo_lite  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
      <link rel="stylesheet" type="text/css" href="../../_static/graphviz.css?v=fd3f3429" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            supremo_lite
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/personalization.html">Personalized Genome Generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/variant_centered_sequences.html">Variant-Centered Sequence Windows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/pam_disruption.html">PAM Disruption Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/prediction_alignment.html">Prediction Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/prediction_alignment_examples.html">Prediction Alignment Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/prediction_alignment_examples.html#variant-type-examples">Variant Type Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/mutagenesis.html">Saturation Mutagenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Notebooks</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/01_getting_started.html">Getting Started with supremo_lite</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/02_personalized_genomes.html">Personalized Genome Generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/03_prediction_alignment.html">Prediction Alignment: From Sequences to Aligned Predictions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/04_pam_disruption.html">PAM Disruption Analysis: Identifying CRISPR-Resistant Variants</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../autoapi/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../conduct.html">Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">supremo_lite</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">supremo_lite.chromosome_utils</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for supremo_lite.chromosome_utils</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Chromosome name matching utilities for supremo_lite.</span>

<span class="sd">This module provides functions for handling mismatches in chromosome naming</span>
<span class="sd">between FASTA references and VCF files using intelligent heuristics.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span>


<div class="viewcode-block" id="ChromosomeMismatchError">
<a class="viewcode-back" href="../../autoapi/supremo_lite/index.html#supremo_lite.chromosome_utils.ChromosomeMismatchError">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">ChromosomeMismatchError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Raised when chromosome names in VCF and reference do not match.</span>

<span class="sd">    This error is raised by default when chromosome names don&#39;t match exactly</span>
<span class="sd">    and automatic chromosome mapping is not enabled.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">pass</span></div>



<div class="viewcode-block" id="normalize_chromosome_name">
<a class="viewcode-back" href="../../autoapi/supremo_lite/index.html#supremo_lite.chromosome_utils.normalize_chromosome_name">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">normalize_chromosome_name</span><span class="p">(</span><span class="n">chrom_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Normalize chromosome name to a standard format.</span>

<span class="sd">    Args:</span>
<span class="sd">        chrom_name: Raw chromosome name from VCF or FASTA</span>

<span class="sd">    Returns:</span>
<span class="sd">        Normalized chromosome name (without &#39;chr&#39; prefix, uppercase)</span>

<span class="sd">    Examples:</span>
<span class="sd">        &#39;chr1&#39; -&gt; &#39;1&#39;</span>
<span class="sd">        &#39;CHR1&#39; -&gt; &#39;1&#39;</span>
<span class="sd">        &#39;chrX&#39; -&gt; &#39;X&#39;</span>
<span class="sd">        &#39;chrMT&#39; -&gt; &#39;MT&#39;</span>
<span class="sd">        &#39;M&#39; -&gt; &#39;MT&#39;  # Mitochondrial normalization</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert to string and strip whitespace</span>
    <span class="n">normalized</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom_name</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="c1"># Remove &#39;chr&#39; prefix (case insensitive)</span>
    <span class="n">normalized</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">normalized</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>

    <span class="c1"># Handle mitochondrial chromosome variants</span>
    <span class="k">if</span> <span class="n">normalized</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;M&quot;</span><span class="p">,</span> <span class="s2">&quot;MITO&quot;</span><span class="p">,</span> <span class="s2">&quot;MITOCHONDRION&quot;</span><span class="p">]:</span>
        <span class="n">normalized</span> <span class="o">=</span> <span class="s2">&quot;MT&quot;</span>

    <span class="c1"># Convert to uppercase for consistency</span>
    <span class="n">normalized</span> <span class="o">=</span> <span class="n">normalized</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">normalized</span></div>



<div class="viewcode-block" id="create_chromosome_mapping">
<a class="viewcode-back" href="../../autoapi/supremo_lite/index.html#supremo_lite.chromosome_utils.create_chromosome_mapping">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_chromosome_mapping</span><span class="p">(</span>
    <span class="n">reference_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">vcf_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a mapping from VCF chromosome names to reference chromosome names.</span>

<span class="sd">    This function uses heuristics to match chromosome names between VCF and FASTA:</span>
<span class="sd">    1. Exact match (case sensitive)</span>
<span class="sd">    2. Exact match (case insensitive)</span>
<span class="sd">    3. Normalized match (with/without &#39;chr&#39; prefix)</span>
<span class="sd">    4. Special cases for mitochondrial chromosomes</span>

<span class="sd">    Args:</span>
<span class="sd">        reference_chroms: Set of chromosome names from reference FASTA</span>
<span class="sd">        vcf_chroms: Set of chromosome names from VCF file</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple of (mapping dict, unmatched set)</span>

<span class="sd">    Example:</span>
<span class="sd">        reference_chroms = {&#39;1&#39;, &#39;2&#39;, &#39;X&#39;, &#39;Y&#39;, &#39;MT&#39;}</span>
<span class="sd">        vcf_chroms = {&#39;chr1&#39;, &#39;chr2&#39;, &#39;chrX&#39;, &#39;chrY&#39;, &#39;chrM&#39;}</span>
<span class="sd">        Returns: {&#39;chr1&#39;: &#39;1&#39;, &#39;chr2&#39;: &#39;2&#39;, &#39;chrX&#39;: &#39;X&#39;, &#39;chrY&#39;: &#39;Y&#39;, &#39;chrM&#39;: &#39;MT&#39;}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mapping</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">unmatched_vcf</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># Try to match each VCF chromosome</span>
    <span class="k">for</span> <span class="n">vcf_chrom</span> <span class="ow">in</span> <span class="n">vcf_chroms</span><span class="p">:</span>
        <span class="n">matched_ref</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># 1. Try exact match (case sensitive)</span>
        <span class="k">if</span> <span class="n">vcf_chrom</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
            <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">vcf_chrom</span>

        <span class="c1"># 2. Try exact match (case insensitive)</span>
        <span class="k">if</span> <span class="n">matched_ref</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">ref_chrom</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">vcf_chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">ref_chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
                    <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">ref_chrom</span>
                    <span class="k">break</span>

        <span class="c1"># 3. Try removing/adding chr prefix</span>
        <span class="k">if</span> <span class="n">matched_ref</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># If VCF has &#39;chr&#39; prefix, try without it</span>
            <span class="k">if</span> <span class="n">vcf_chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">):</span>
                <span class="n">no_chr</span> <span class="o">=</span> <span class="n">vcf_chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">no_chr</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                    <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">no_chr</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Try case insensitive match without chr</span>
                    <span class="k">for</span> <span class="n">ref_chrom</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">no_chr</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">ref_chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
                            <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">ref_chrom</span>
                            <span class="k">break</span>

            <span class="c1"># If VCF doesn&#39;t have &#39;chr&#39; prefix, try with it</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">with_chr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;chr</span><span class="si">{</span><span class="n">vcf_chrom</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">with_chr</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                    <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">with_chr</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Try case insensitive match with chr</span>
                    <span class="k">for</span> <span class="n">ref_chrom</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">with_chr</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">ref_chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
                            <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">ref_chrom</span>
                            <span class="k">break</span>

        <span class="c1"># 4. Try normalized matching (handles mitochondrial variants)</span>
        <span class="k">if</span> <span class="n">matched_ref</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">vcf_normalized</span> <span class="o">=</span> <span class="n">normalize_chromosome_name</span><span class="p">(</span><span class="n">vcf_chrom</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">ref_chrom</span> <span class="ow">in</span> <span class="n">reference_chroms</span><span class="p">:</span>
                <span class="n">ref_normalized</span> <span class="o">=</span> <span class="n">normalize_chromosome_name</span><span class="p">(</span><span class="n">ref_chrom</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">vcf_normalized</span> <span class="o">==</span> <span class="n">ref_normalized</span><span class="p">:</span>
                    <span class="n">matched_ref</span> <span class="o">=</span> <span class="n">ref_chrom</span>
                    <span class="k">break</span>

        <span class="c1"># Record result</span>
        <span class="k">if</span> <span class="n">matched_ref</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">mapping</span><span class="p">[</span><span class="n">vcf_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">matched_ref</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">unmatched_vcf</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">vcf_chrom</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">unmatched_vcf</span></div>



<div class="viewcode-block" id="apply_chromosome_mapping">
<a class="viewcode-back" href="../../autoapi/supremo_lite/chromosome_utils/index.html#supremo_lite.chromosome_utils.apply_chromosome_mapping">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">apply_chromosome_mapping</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">mapping</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Apply chromosome name mapping to a variants DataFrame.</span>

<span class="sd">    Args:</span>
<span class="sd">        variants_df: Pandas DataFrame with &#39;chrom&#39; column</span>
<span class="sd">        mapping: Dictionary mapping original to new chromosome names</span>

<span class="sd">    Returns:</span>
<span class="sd">        Modified DataFrame with updated chromosome names</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variants_df</span> <span class="o">=</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Apply mapping to chromosome column</span>
    <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mapping</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">variants_df</span></div>



<div class="viewcode-block" id="get_chromosome_match_report">
<a class="viewcode-back" href="../../autoapi/supremo_lite/chromosome_utils/index.html#supremo_lite.chromosome_utils.get_chromosome_match_report">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_chromosome_match_report</span><span class="p">(</span>
    <span class="n">reference_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">vcf_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">mapping</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">unmatched</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a human-readable report of chromosome matching results.</span>

<span class="sd">    Args:</span>
<span class="sd">        reference_chroms: Set of reference chromosome names</span>
<span class="sd">        vcf_chroms: Set of VCF chromosome names</span>
<span class="sd">        mapping: Successful mappings</span>
<span class="sd">        unmatched: Unmatched VCF chromosomes</span>

<span class="sd">    Returns:</span>
<span class="sd">        Formatted report string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">report_lines</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Chromosome Matching Report&quot;</span><span class="p">)</span>
    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;=&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>

    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Reference chromosomes (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">reference_chroms</span><span class="p">)</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">reference_chroms</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;VCF chromosomes (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">)</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">mapping</span><span class="p">:</span>
        <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully matched (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span><span class="si">}</span><span class="s2">):&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">vcf_chrom</span><span class="p">,</span> <span class="n">ref_chrom</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mapping</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">vcf_chrom</span> <span class="o">!=</span> <span class="n">ref_chrom</span><span class="p">:</span>
                <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  &#39;</span><span class="si">{</span><span class="n">vcf_chrom</span><span class="si">}</span><span class="s2">&#39; -&gt; &#39;</span><span class="si">{</span><span class="n">ref_chrom</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  &#39;</span><span class="si">{</span><span class="n">vcf_chrom</span><span class="si">}</span><span class="s2">&#39; (exact match)&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">unmatched</span><span class="p">:</span>
        <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unmatched VCF chromosomes (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unmatched</span><span class="p">)</span><span class="si">}</span><span class="s2">):&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">unmatched</span><span class="p">):</span>
            <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  &#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">&#39; (no suitable reference match found)&quot;</span><span class="p">)</span>

    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">coverage</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">vcf_chroms</span> <span class="k">else</span> <span class="mi">100</span>
    <span class="n">report_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Matching coverage: </span><span class="si">{</span><span class="n">coverage</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">% (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">report_lines</span><span class="p">)</span></div>



<div class="viewcode-block" id="match_chromosomes_with_report">
<a class="viewcode-back" href="../../autoapi/supremo_lite/index.html#supremo_lite.chromosome_utils.match_chromosomes_with_report">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">match_chromosomes_with_report</span><span class="p">(</span>
    <span class="n">reference_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">vcf_chroms</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">auto_map_chromosomes</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Match chromosomes and optionally print a detailed report.</span>

<span class="sd">    Args:</span>
<span class="sd">        reference_chroms: Set of reference chromosome names</span>
<span class="sd">        vcf_chroms: Set of VCF chromosome names</span>
<span class="sd">        verbose: Whether to print matching report</span>
<span class="sd">        auto_map_chromosomes: Whether to automatically map chromosome names when they don&#39;t</span>
<span class="sd">                             match exactly (default: False). When False, raises</span>
<span class="sd">                             ChromosomeMismatchError if names don&#39;t match.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple of (mapping dict, unmatched set)</span>

<span class="sd">    Raises:</span>
<span class="sd">        ChromosomeMismatchError: If auto_map_chromosomes=False and chromosome names don&#39;t</span>
<span class="sd">                                match exactly between VCF and reference</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check for exact matches first</span>
    <span class="n">exact_matches</span> <span class="o">=</span> <span class="n">reference_chroms</span> <span class="o">&amp;</span> <span class="n">vcf_chroms</span>
    <span class="n">needs_mapping</span> <span class="o">=</span> <span class="n">vcf_chroms</span> <span class="o">-</span> <span class="n">exact_matches</span>

    <span class="c1"># If all chromosomes match exactly, no mapping needed</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">needs_mapping</span><span class="p">:</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="p">{</span><span class="n">chrom</span><span class="p">:</span> <span class="n">chrom</span> <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">vcf_chroms</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">mapping</span><span class="p">,</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># If mapping is needed but not enabled, raise error</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">auto_map_chromosomes</span><span class="p">:</span>
        <span class="c1"># Format chromosome lists for error message</span>
        <span class="n">vcf_list</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">))</span>
        <span class="n">ref_list</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">reference_chroms</span><span class="p">))</span>

        <span class="n">error_msg</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Chromosome names in VCF and reference do not match.</span><span class="se">\n\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;VCF chromosomes: </span><span class="si">{</span><span class="n">vcf_list</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Reference chromosomes: </span><span class="si">{</span><span class="n">ref_list</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;To enable automatic chromosome name mapping, add auto_map_chromosomes=True:</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  get_personal_genome(..., auto_map_chromosomes=True)</span><span class="se">\n\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Alternatively, rename chromosomes in your VCF to match the reference.&quot;</span>
        <span class="p">)</span>
        <span class="k">raise</span> <span class="n">ChromosomeMismatchError</span><span class="p">(</span><span class="n">error_msg</span><span class="p">)</span>

    <span class="c1"># Automatic mapping is enabled - use heuristics</span>
    <span class="n">mapping</span><span class="p">,</span> <span class="n">unmatched</span> <span class="o">=</span> <span class="n">create_chromosome_mapping</span><span class="p">(</span><span class="n">reference_chroms</span><span class="p">,</span> <span class="n">vcf_chroms</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_chroms</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">any</span><span class="p">(</span><span class="n">k</span> <span class="o">!=</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">mapping</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
    <span class="p">):</span>
        <span class="n">report</span> <span class="o">=</span> <span class="n">get_chromosome_match_report</span><span class="p">(</span>
            <span class="n">reference_chroms</span><span class="p">,</span> <span class="n">vcf_chroms</span><span class="p">,</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">unmatched</span>
        <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">report</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">unmatched</span><span class="p">:</span>
        <span class="n">chrom_list</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">unmatched</span><span class="p">))</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Skipped </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unmatched</span><span class="p">)</span><span class="si">}</span><span class="s2"> chromosome(s) not in reference: </span><span class="si">{</span><span class="n">chrom_list</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">unmatched</span></div>



<div class="viewcode-block" id="validate_chromosomes_early">
<a class="viewcode-back" href="../../autoapi/supremo_lite/chromosome_utils/index.html#supremo_lite.chromosome_utils.validate_chromosomes_early">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">validate_chromosomes_early</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">variants_fn</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Efficiently validate chromosome compatibility before loading all variant data.</span>

<span class="sd">    This function optimizes chromosome checking by:</span>
<span class="sd">    - For VCF file paths: Reading only the chromosome column (very fast and memory efficient)</span>
<span class="sd">    - For DataFrames: Using existing data without reloading</span>
<span class="sd">    - Returning chromosome sets for reuse in subsequent mapping operations</span>

<span class="sd">    Args:</span>
<span class="sd">        reference: Reference genome (dict-like object with .keys())</span>
<span class="sd">        variants_fn: VCF file path (str) or DataFrame with variants</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple of (ref_chroms, vcf_chroms) as sets</span>

<span class="sd">    Note:</span>
<span class="sd">        This function does NOT raise errors or perform mapping - it only extracts</span>
<span class="sd">        chromosome names efficiently. Use match_chromosomes_with_report() for</span>
<span class="sd">        actual validation and mapping.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.variant_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_vcf_chromosomes</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

    <span class="c1"># Get reference chromosomes</span>
    <span class="n">ref_chroms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="c1"># Get VCF chromosomes efficiently based on input type</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">variants_fn</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="c1"># VCF file path - use efficient chromosome extraction (reads only first column)</span>
        <span class="n">vcf_chroms</span> <span class="o">=</span> <span class="n">get_vcf_chromosomes</span><span class="p">(</span><span class="n">variants_fn</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">variants_fn</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="c1"># DataFrame - extract unique chromosome names from chrom column</span>
        <span class="n">vcf_chroms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">variants_fn</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Other formats - try to get chrom column</span>
        <span class="n">vcf_chroms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">variants_fn</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">ref_chroms</span><span class="p">,</span> <span class="n">vcf_chroms</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Gladstone Institutes.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>