

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Personalized Genome Generation &mdash; supremo_lite  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Prediction Alignment: From Sequences to Aligned Predictions" href="03_prediction_alignment.html" />
    <link rel="prev" title="Getting Started with supremo_lite" href="01_getting_started.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            supremo_lite
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/personalization.html">Personalized Genome Generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/variant_centered_sequences.html">Variant-Centered Sequence Windows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/pam_disruption.html">PAM Disruption Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/prediction_alignment.html">Prediction Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/prediction_alignment_examples.html">Prediction Alignment Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/prediction_alignment_examples.html#variant-type-examples">Variant Type Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/mutagenesis.html">Saturation Mutagenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Notebooks</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="01_getting_started.html">Getting Started with supremo_lite</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Personalized Genome Generation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#learning-objectives">Learning Objectives</a></li>
<li class="toctree-l2"><a class="reference internal" href="#setup">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#understanding-variant-types">Understanding Variant Types</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basic-genome-personalization">Basic Genome Personalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examining-variant-application">Examining Variant Application</a></li>
<li class="toctree-l2"><a class="reference internal" href="#skipped-variants-and-overlap-detection">Skipped Variants and Overlap Detection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-length-changes-from-indels">Sequence Length Changes from Indels</a></li>
<li class="toctree-l2"><a class="reference internal" href="#best-practices">Best Practices</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#always-use-verbose-mode-during-development">1. <strong>Always use verbose mode during development</strong></a></li>
<li class="toctree-l3"><a class="reference internal" href="#use-chunked-processing-for-large-vcf-files">2. <strong>Use chunked processing for large VCF files</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#next-steps">Next Steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="03_prediction_alignment.html">Prediction Alignment: From Sequences to Aligned Predictions</a></li>
<li class="toctree-l1"><a class="reference internal" href="04_pam_disruption.html">PAM Disruption Analysis: Identifying CRISPR-Resistant Variants</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../autoapi/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conduct.html">Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">supremo_lite</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Personalized Genome Generation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/02_personalized_genomes.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="personalized-genome-generation">
<h1>Personalized Genome Generation<a class="headerlink" href="#personalized-genome-generation" title="Link to this heading"></a></h1>
<p>Comprehensive guide to creating personalized genomes using supremo_lite.</p>
<section id="learning-objectives">
<h2>Learning Objectives<a class="headerlink" href="#learning-objectives" title="Link to this heading"></a></h2>
<p>By the end of this notebook, you will understand:</p>
<ul class="simple">
<li><p>How different variant types (SNV, INS, DEL, MNV) are applied</p></li>
<li><p>Chromosome order preservation in outputs</p></li>
<li><p>Verbose mode for debugging variant application</p></li>
<li><p>Skipped variant reporting</p></li>
<li><p>Chunked processing for large VCF files</p></li>
<li><p>Best practices for genome personalization</p></li>
</ul>
</section>
<section id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Link to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">supremo_lite</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pyfaidx</span><span class="w"> </span><span class="kn">import</span> <span class="n">Fasta</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="c1"># Set up paths to test data</span>
<span class="n">test_data_dir</span> <span class="o">=</span> <span class="s2">&quot;../../tests/data&quot;</span>
<span class="n">reference_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">test_data_dir</span><span class="p">,</span> <span class="s2">&quot;test_genome.fa&quot;</span><span class="p">)</span>
<span class="n">multi_vcf_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">test_data_dir</span><span class="p">,</span> <span class="s2">&quot;multi&quot;</span><span class="p">,</span> <span class="s2">&quot;multi.vcf&quot;</span><span class="p">)</span>

<span class="c1"># Load reference genome</span>
<span class="n">reference</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">reference_path</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;supremo_lite version: </span><span class="si">{</span><span class="n">sl</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference genome loaded: </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>supremo_lite version: 0.5.4
Reference genome loaded: [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;, &#39;chr6&#39;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="understanding-variant-types">
<h2>Understanding Variant Types<a class="headerlink" href="#understanding-variant-types" title="Link to this heading"></a></h2>
<p>supremo_lite automatically classifies variants, for more information on how this classification is done and which VCF representations are supported, see the variant classification flowchart.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load variants</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="n">multi_vcf_path</span><span class="p">,</span> <span class="n">classify_variants</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Variants in VCF:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variants in VCF:
  chrom  pos1 id            ref       alt    info  vcf_line variant_type
0  chr1    10  .              A         T  DP=100         9          SNV
1  chr1    11  .              T   TATATAG  DP=100        10          INS
2  chr1    11  .  TATTTTCGAGAAT         T  DP=100        11          DEL
3  chr1    11  .              T    TGGGGG  DP=100        12          INS
4  chr1    25  .              A  AATATAGG  DP=100        13          INS
5  chr1    50  .              G    GGGGGG  DP=100        14          INS
6  chr1    52  .         GTTTTA         G  DP=100        15          DEL
7  chr1    76  .          ATATT     AGATT  DP=100        16          MNV
</pre></div>
</div>
</div>
</div>
</section>
<section id="basic-genome-personalization">
<h2>Basic Genome Personalization<a class="headerlink" href="#basic-genome-personalization" title="Link to this heading"></a></h2>
<p>Let’s create a personalized genome and examine the results:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create personalized genome with verbose output</span>
<span class="n">personal_genome</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">get_personal_genome</span><span class="p">(</span>
    <span class="n">reference_fn</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
    <span class="n">variants_fn</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
    <span class="n">encode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Get string sequences for easy comparison</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Generated personalized genome:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">personal_genome</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">personal_genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">])</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated personalized genome:
chr1: 93 bp
chr2: 80 bp
chr3: 80 bp
chr4: 80 bp
chr5: 80 bp
chr6: 80 bp
</pre></div>
</div>
</div>
</div>
</section>
<section id="examining-variant-application">
<h2>Examining Variant Application<a class="headerlink" href="#examining-variant-application" title="Link to this heading"></a></h2>
<p>Let’s look at specific regions to see how each variant type was applied:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to show variant context</span>
<span class="k">def</span><span class="w"> </span><span class="nf">show_variant_context</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos1</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">personal_genome</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Display reference and personalized sequence around a variant.&quot;&quot;&quot;</span>
    <span class="c1"># VCF uses 1-based, Python uses 0-based</span>
    <span class="n">pos0</span> <span class="o">=</span> <span class="n">pos1</span> <span class="o">-</span> <span class="mi">1</span>

    <span class="c1"># Get reference context</span>
    <span class="n">ref_start</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pos0</span> <span class="o">-</span> <span class="n">window</span><span class="p">)</span>
    <span class="n">ref_end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="p">[</span><span class="n">chrom</span><span class="p">]),</span> <span class="n">pos0</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">+</span> <span class="n">window</span><span class="p">)</span>
    <span class="n">ref_context</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">ref_start</span><span class="p">:</span><span class="n">ref_end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>

    <span class="c1"># Calculate where variant appears in context</span>
    <span class="n">var_offset</span> <span class="o">=</span> <span class="n">pos0</span> <span class="o">-</span> <span class="n">ref_start</span>

    <span class="c1"># Get personalized context (may have different length due to indel)</span>
    <span class="c1"># For personalized, we need to account for the length change</span>
    <span class="n">len_diff</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
    <span class="n">pers_end</span> <span class="o">=</span> <span class="n">ref_end</span> <span class="o">+</span> <span class="n">len_diff</span>
    <span class="n">pers_context</span> <span class="o">=</span> <span class="n">personal_genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">ref_start</span><span class="p">:</span><span class="n">pers_end</span><span class="p">]</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">pos1</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ref</span><span class="si">}</span><span class="s2"> → </span><span class="si">{</span><span class="n">alt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Reference:    </span><span class="si">{</span><span class="n">ref_context</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;              </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">var_offset</span><span class="si">}{</span><span class="s1">&#39;^&#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Personalized: </span><span class="si">{</span><span class="n">pers_context</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;              </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">var_offset</span><span class="si">}{</span><span class="s1">&#39;^&#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">()</span>


<span class="c1"># Show examples of each variant type</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=&quot;</span> <span class="o">*</span> <span class="mi">60</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;VARIANT APPLICATION EXAMPLES&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=&quot;</span> <span class="o">*</span> <span class="mi">60</span><span class="p">)</span>

<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>  <span class="c1"># Show first 4 variants</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;variant_type&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> variant:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">60</span><span class="p">)</span>
    <span class="n">show_variant_context</span><span class="p">(</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;pos1&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">],</span> <span class="n">reference</span><span class="p">,</span> <span class="n">personal_genome</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>============================================================
VARIANT APPLICATION EXAMPLES
============================================================

SNV variant:
------------------------------------------------------------
chr1:10 A → T

Reference:    ATGAATATAATATTTTCGAGAATTACTCCT
                       ^
Personalized: ATGAATATATTATATAGATTTTCGAGAATT
                       ^


INS variant:
------------------------------------------------------------
chr1:11 T → TATATAG

Reference:    ATGAATATAATATTTTCGAGAATTACTCCTT
                        ^
Personalized: ATGAATATATTATATAGATTTTCGAGAATTAATATAG
                        ^^^^^^^


DEL variant:
------------------------------------------------------------
chr1:11 TATTTTCGAGAAT → T

Reference:    ATGAATATAATATTTTCGAGAATTACTCCTTTTGGAAATGGAA
                        ^^^^^^^^^^^^^
Personalized: ATGAATATATTATATAGATTTTCGAGAATTA
                        ^


INS variant:
------------------------------------------------------------
chr1:11 T → TGGGGG

Reference:    ATGAATATAATATTTTCGAGAATTACTCCTT
                        ^
Personalized: ATGAATATATTATATAGATTTTCGAGAATTAATATA
                        ^^^^^^
</pre></div>
</div>
</div>
</div>
</section>
<section id="skipped-variants-and-overlap-detection">
<h2>Skipped Variants and Overlap Detection<a class="headerlink" href="#skipped-variants-and-overlap-detection" title="Link to this heading"></a></h2>
<p>When variants overlap or can’t be applied, supremo_lite tracks and reports them in verbose mode. Let’s see this in action:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The multi.vcf file contains overlapping variants</span>

<span class="c1"># Create personalized genome with verbose output</span>
<span class="n">personal_genome</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">get_personal_genome</span><span class="p">(</span>
    <span class="n">reference_fn</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
    <span class="n">variants_fn</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
    <span class="n">encode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Get string sequences for easy comparison</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Show detailed processing information</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>🧬 Processing 8 variants across 1 chromosomes
   Phase 1: 8 standard variants (SNV, MNV, INS, DEL, SV_DUP, SV_INV)
   Phase 2: 0 BND variants for semantic classification
🔄 Processing chromosome chr1: 8 variants (4 INS, 2 DEL, 1 SNV, 1 MNV)
  ⚠️  Skipped 2 variant(s):
     • overlap with previous variant: VCF line(s) 11, 12 at chr1:11
  ✅ Applied 6/8 variants (2 skipped)
🧬 Completed: 8/8 variants processed → 6 sequences
</pre></div>
</div>
</div>
</div>
</section>
<section id="sequence-length-changes-from-indels">
<h2>Sequence Length Changes from Indels<a class="headerlink" href="#sequence-length-changes-from-indels" title="Link to this heading"></a></h2>
<p>Insertions and deletions change the length of chromosomes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate expected length changes</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosome length analysis:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=&quot;</span> <span class="o">*</span> <span class="mi">60</span><span class="p">)</span>

<span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">reference</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">ref_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="p">[</span><span class="n">chrom</span><span class="p">])</span>
    <span class="n">pers_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">personal_genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">])</span>

    <span class="c1"># Calculate total length change from variants</span>
    <span class="n">chrom_variants</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="n">variants</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">]</span>
    <span class="n">total_change</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">])</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">chrom_variants</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Note: total_change may differ from actual due to overlapping variants</span>
    <span class="n">actual_change</span> <span class="o">=</span> <span class="n">pers_len</span> <span class="o">-</span> <span class="n">ref_len</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Reference length:    </span><span class="si">{</span><span class="n">ref_len</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Personalized length: </span><span class="si">{</span><span class="n">pers_len</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Actual change:       </span><span class="si">{</span><span class="n">actual_change</span><span class="si">:</span><span class="s2">+,</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Expected change*:    </span><span class="si">{</span><span class="n">total_change</span><span class="si">:</span><span class="s2">+,</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Variants on chrom:   </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom_variants</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">total_change</span> <span class="o">!=</span> <span class="n">actual_change</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  ⚠️  Difference due to overlapping/skipped variants&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Chromosome length analysis:
============================================================

chr1:
  Reference length:    80 bp
  Personalized length: 93 bp
  Actual change:       +13 bp
  Expected change*:    +6 bp
  Variants on chrom:   8
  ⚠️  Difference due to overlapping/skipped variants

chr2:
  Reference length:    80 bp
  Personalized length: 80 bp
  Actual change:       +0 bp
  Expected change*:    +0 bp
  Variants on chrom:   0

chr3:
  Reference length:    80 bp
  Personalized length: 80 bp
  Actual change:       +0 bp
  Expected change*:    +0 bp
  Variants on chrom:   0

chr4:
  Reference length:    80 bp
  Personalized length: 80 bp
  Actual change:       +0 bp
  Expected change*:    +0 bp
  Variants on chrom:   0

chr5:
  Reference length:    80 bp
  Personalized length: 80 bp
  Actual change:       +0 bp
  Expected change*:    +0 bp
  Variants on chrom:   0

chr6:
  Reference length:    80 bp
  Personalized length: 80 bp
  Actual change:       +0 bp
  Expected change*:    +0 bp
  Variants on chrom:   0
</pre></div>
</div>
</div>
</div>
</section>
<section id="best-practices">
<h2>Best Practices<a class="headerlink" href="#best-practices" title="Link to this heading"></a></h2>
<section id="always-use-verbose-mode-during-development">
<h3>1. <strong>Always use verbose mode during development</strong><a class="headerlink" href="#always-use-verbose-mode-during-development" title="Link to this heading"></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">personal_genome</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">get_personal_genome</span><span class="p">(</span>
    <span class="n">reference_fn</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
    <span class="n">variants_fn</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>  <span class="c1"># See what&#39;s happening!</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="use-chunked-processing-for-large-vcf-files">
<h3>2. <strong>Use chunked processing for large VCF files</strong><a class="headerlink" href="#use-chunked-processing-for-large-vcf-files" title="Link to this heading"></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">personal_genome</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">get_personal_genome</span><span class="p">(</span>
    <span class="n">reference_fn</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
    <span class="n">variants_fn</span><span class="o">=</span><span class="s1">&#39;large_variants.vcf&#39;</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="o">=</span><span class="mi">10000</span>  <span class="c1"># Process vcf in 10k chunks (default is 1)</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="next-steps">
<h2>Next Steps<a class="headerlink" href="#next-steps" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><strong><a class="reference internal" href="03_prediction_alignment.html"><span class="std std-doc">03_prediction_alignment.ipynb</span></a></strong> - Generate variant centered window sequences and align model predictions</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="01_getting_started.html" class="btn btn-neutral float-left" title="Getting Started with supremo_lite" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="03_prediction_alignment.html" class="btn btn-neutral float-right" title="Prediction Alignment: From Sequences to Aligned Predictions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Gladstone Institutes.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>